Expanded objects

pnearing opened the pull request:
I expanded TiledObject to recognize ellipses, text and points.  I added a new property to the object called object_type which is meant to be a literal string, currently it will be one of "rectangle", "text" "polygon" "polyline", "ellipse", and "point".

Also my IDE cleaned up and made some date changes.

pnearing committed:
Update Copyright Years and Object Types

Changed the copyright years in multiple files for consistency. Adjusted object types in `pytmx.py` to include additional types such as ellipse, point, and text, along with their respective attributes. Refined imports and typing hints across various modules for clarity and efficiency.

pnearing committed:
Adopt consistent string quoting in pytmx module

Changed all single quotes to double quotes for string literals in pytmx.py. This ensures uniformity in the codebase and improves readability and maintainability of the code. No functional changes were made.

joereynolds commented:
Hmm, those changes made by your IDE are a bit suspect

They've changed the copyright years to be in the past. What's worse though is that they've removed the return typehints which are useful for people working with PyTMX

pnearing commented:
I didn't notice the type hint changes.  I'll go through them and check it out, I'm all about typing.

pnearing committed:
Revert "Update Copyright Years and Object Types"

This reverts commit ec662ae9

pnearing committed:
Add get_tile_flags_by_gid method to pytmx.py

Introduce a new method, get_tile_flags_by_gid, to fetch tile flags using the global tile ID (GID). This improves the convenience of accessing tile properties directly by their GID.

pnearing commented:
Okay, I've removed the changes my IDE made, and left only the object changes, also I added a function called get_tile_flags_by_gid to get the flags of a gid, I use this for correctly flipping and rotating my colliders.

zommerfelds commented on pytmx/pytmx.py:
```diff
@@ -1022,6 +1023,15 @@ def get_tile_colliders(self) -> Iterable[tuple[int, list[dict]]]:
+            if gid == tile_gid:
+                return flags
+
+
```
Two empty lines too many?

pnearing replied:
Fixed

zommerfelds commented on pytmx/pytmx.py:
```diff
@@ -1103,6 +1113,7 @@ def register_gid(
             flags = TileFlags(0, 0, 0)
 
         if tiled_gid:
+
```
Unnecessary change?

pnearing replied:
Fixed, all these newlines, was me understanding the code, lol

zommerfelds commented on pytmx/pytmx.py:
```diff
@@ -1550,14 +1561,43 @@ def read_points(text):
             self.closed = False
 
+        ellipse = node.find("ellipse")
+        if ellipse is not None:
```
Newline here for consistency?

zommerfelds commented on pytmx/pytmx.py:
```diff
@@ -1550,14 +1561,43 @@ def read_points(text):
+        if text is not None:
+            self.object_type = "text"
+            # NOTE: The defaults have been taken from the tiled editor version 1.11.0
+            setattr(self, "text", text.text)
```
Why use `setattr` if you could simply do `self.text = text.text`

pnearing replied:
This one I was thinking of keeping in line with the usage of setattr across the code base, but looking at it now it seems a little overkill, and if I do it for text, I should do it for all the properties.

pnearing replied:
Done.

zommerfelds commented on pytmx/util_pygame_sdl2.py:
```diff
@@ -16,6 +16,7 @@
 You should have received a copy of the GNU Lesser General Public
 License along with pytmx.  If not, see <http://www.gnu.org/licenses/>.
 """
+
```
Unnecessary change?

pnearing replied:
Fixed

zommerfelds reviewed (commented):
Leaving some drive-by comments :)

I was about to make the same changes, so let's try to get this PR in, it's useful!

zommerfelds commented:
BTW I switched to using @pnearing's repo for my project, and it does work.

One thing to note is that images/tiles are considered rectangles. Which is not wrong, as you can use the `gid` property to distinguish, but we could consider setting the `object_type` to `tile`.

zommerfelds commented:
Sorry to add yet another comment, but another thing that would be convenient is if rectangles also had `points` generated... (future follow-up)
Meanwhile I'll have to manually convert from rectangle to points, so I can draw using the same polygon function.

pnearing committed:
Update pytmx.py

Remove unnecessary changes

pnearing committed:
Update util_pygame_sdl2.py

Remove unnecessary change

pnearing committed:
Move from using setattr to setting the properties directly.

pnearing committed:
feat: add points calculation for rectangle objects in TMX parser

pnearing commented:
> Sorry to add yet another comment, but another thing that would be convenient is if rectangles also had `points` generated... (future follow-up) Meanwhile I'll have to manually convert from rectangle to points, so I can draw using the same polygon function.

If you happen to still be using my repo, I added points for rectangles.

pnearing committed:
feat: add TiledGroupLayer to public API exports

JaskRendix commented:
@pnearing hey man, since you're working on that section already, would you mind swapping this piece out while you're in there?
```
        if points:
            x1 = x2 = y1 = y2 = 0
            for x, y in points:
                if x < x1:
                    x1 = x
                if x > x2:
                    x2 = x
                if y < y1:
                    y1 = y
                if y > y2:
                    y2 = y
            self.width = abs(x1) + abs(x2)
            self.height = abs(y1) + abs(y2)
```
with this
```
        if points:
            xs, ys = zip(*points)
            self.width = max(xs) - min(xs)
            self.height = max(ys) - min(ys)
```

pnearing committed:
refactor: simplify polygon bounds calculation using zip and min/max

pnearing commented:
Done.

pnearing committed:
feat: add rotation handling for tilemap objects based on GID flags

pnearing committed:
ci: remove GitHub Actions test workflow configuration

pnearing committed:
chore: add .venv directory to gitignore

pnearing commented:
To anyone who may be using my version of pytmx, I forked the project as pytmx-ng and will be happy to maintain it.  the new code can be found here: https://github.com/pnearing/pytmx-ng
